x/dclupgrade/handler_test.go

```diff
@@ -0,0 +1,250 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//nolint:testpackage,lll
+package dclupgrade
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/testutil/testdata"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	testconstants "github.com/zigbee-alliance/distributed-compliance-ledger/integration_tests/constants"
+	testkeeper "github.com/zigbee-alliance/distributed-compliance-ledger/testutil/keeper"
+	dclauthtypes "github.com/zigbee-alliance/distributed-compliance-ledger/x/dclauth/types"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/keeper"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/types"
+)
+
+type TestSetup struct {
+	T *testing.T
+	// Cdc         *amino.Codec
+	Ctx           sdk.Context
+	Wctx          context.Context
+	Keeper        *keeper.Keeper
+	DclauthKeeper *DclauthKeeperMock
+	UpgradeKeeper *UpgradeKeeperMock
+	Handler       sdk.Handler
+	// Querier     sdk.Querier
+	Vendor   sdk.AccAddress
+	VendorID int32
+	Plan     types.Plan
+}
+
+func (setup *TestSetup) AddAccount(
+	accAddress sdk.AccAddress,
+	roles []dclauthtypes.AccountRole,
+	vendorID int32,
+) {
+	for _, role := range roles {
+		setup.DclauthKeeper.On("HasRole", mock.Anything, accAddress, role).Return(true)
+	}
+	setup.DclauthKeeper.On("HasRole", mock.Anything, accAddress, mock.Anything).Return(false)
+
+	setup.DclauthKeeper.On("HasVendorID", mock.Anything, accAddress, vendorID).Return(true)
+	setup.DclauthKeeper.On("HasVendorID", mock.Anything, accAddress, mock.Anything).Return(false)
+}
+
+func (setup *TestSetup) ScheduleUpgrade(
+	plan types.Plan,
+) {
+	setup.UpgradeKeeper.On("ScheduleUpgrade", mock.Anything, plan).Return(nil)
+}
+
+func Setup(t *testing.T) TestSetup {
+	dclauthKeeper := &DclauthKeeperMock{}
+	upgradeKeeper := &UpgradeKeeperMock{}
+	keeper, ctx := testkeeper.DclupgradeKeeper(t, dclauthKeeper, upgradeKeeper)
+
+	vendor := GenerateAccAddress()
+	vendorID := testconstants.VendorID1
+	plan := testconstants.Plan
+	setup := TestSetup{
+		T:             t,
+		Ctx:           ctx,
+		Wctx:          sdk.WrapSDKContext(ctx),
+		Keeper:        keeper,
+		DclauthKeeper: dclauthKeeper,
+		UpgradeKeeper: upgradeKeeper,
+		Handler:       NewHandler(*keeper),
+		Vendor:        vendor,
+		VendorID:      vendorID,
+		Plan:          plan,
+	}
+
+	setup.AddAccount(vendor, []dclauthtypes.AccountRole{types.UpgradeApprovalRole}, vendorID)
```

Comment:
It would be better to create all the accounts and specify corresponding behavior of `DclauthKeeper.CountAccountsWithRole` right in a test to make it clear how the quorum number is determined in this specific test.

Reply:
Do not rely in tests on the constant `UpgradeApprovalRole` from the module being tested. Instead explicitly specify the concrete role `Trustee`.

Reply:
fixed